### ISSUE TITLE
Configurable development port for local webserver start

### ISSUE THREAD

**Issue by scarcloud** [symfony-cli/symfony-cli#42]
Wenn using `symfony serve`, you can configure the application port, but the development server port is always randomized.

Is there an easy way to implement a `--dev-port` parameter keeps randomizing the development port when omitted?

**Comment by cyildirim** [comment]
@scarcloud  you can use `--port` flag.

 `symfony serve --port 2000`

**Comment by fabpot** [comment]
What is the development port?

**Comment by scarcloud** [comment]
> @scarcloud you can use `--port` flag.
> 
> `symfony serve --port 2000`

@cyildirim That's the parameter for the **web** server port, but symfony also starts a **development** server:
```
user@machine:/current/path$ symfony serve --port 2000
Tailing Web Server log file (/home/user/.symfony/log/fd7ccd6ae346f36b24b88135fc7cf96c353e4236.log)
Tailing PHP log file (/home/user/.symfony/log/fd7ccd6ae346f36b24b88135fc7cf96c353e4236/7daf403c7589f4927632ed3b6af762a992f09b78.log)
                                                                                                                        
 [OK] Web server listening                                                                                              
      The Web server is using PHP CLI 8.0.14                                                                            
      https://127.0.0.1:2000                                                                                            
                                                                                                                        

[Web Server ] Jan 13 14:46:52 |DEBUG  | PHP    Reloading PHP versions
[Web Server ] Jan 13 14:46:52 |DEBUG  | PHP    Using PHP version 8.0.14 (from .php-version from current dir: /current/path/.php-version)
[Web Server ] Jan 13 14:46:52 |INFO   | PHP    listening path="/usr/bin/php8.0" php="8.0.14" port=62470
[PHP        ] [Thu Jan 13 14:46:53 2022] PHP 8.0.14 Development Server (http://127.0.0.1:62470) started
```
I'm talking about the last two lines, where the port 62470 is defined. Maybe I just use a wrong term.

**Comment by fabpot** [comment]
This is the port of the PHP-FPM/PHP CGI/PHP CLI (depending on what you have) process. Why would you want to configure it to something specific? You should not interact with it.

**Comment by scarcloud** [comment]
> Why would you want to configure it to something specific?

I use PHPStorm as my IDE and want to utilize step-debugging via XDebug. When configuring the server (File > Settings > PHP > Servers) I can enter the port to listen to, but step-debugging only works if I specify the development port.
However, because the development port changes with each instance of `symfony serve`, I have to manually adjust it in the settings, else PHPStorm suggets creating a "new" server configuration with the current development port.

It's a convenience thing, tbh. Maybe I got my PHP / XDebug / PHPStorm configuration wrong, I'm not quite sure.
I've observed that when using PHP-FPM, I can use the web server port, but using WSL 1 my logs are getting spammed with errors.

**Comment by LeeSaferite** [comment]
I have this same issue FWIW. The random port for the worker is the port that PhpStorm sees when mapping the incoming debugger request. This means that each time I run `symfony server:start` I either have to reconfigure path mapping or edit the server config to change the port. Having a way to set the the worker port would be useful in this scenario. Better would be fixing it so that PhpStorm sees the main server port and not the worker port.

**Comment by Jeto143** [comment]
I'm using Symfony 6 w/ PHP 8.1 (through Windows + WSL2, if that makes any difference) and I have the same issue, plus I actually *need* to reach that random port with my browser to access the app (which means I can't even bookmark it). If I try to access it with the web server port I get an "Unable to connect" error.

How to solve this?

**Comment by antonymous** [comment]
I solved this with Zero Config Debugging](https://www.jetbrains.com/help/phpstorm/zero-configuration-debugging.html).
Be sure to uncheck "Ignore external connections through unregistered server configurations" and IDE will ask you to accept incoming connection from any new worker's port.

**Comment by scarcloud** [comment]
> But I guess Symfony CLI can do nothing here until such port fixing is supported by PHP, right?
@antonymous I haven't thoroughly checked the source code, so I don't really know how complicated it would be to provide an `--dev-port` argument. From my point of view, it would be the best solution though.

**Comment by ah3nan** [comment]
same issue here

**Comment by Nsbx** [comment]
For strange reason (the reason was me 100% me).
I encountered the same problem, the server was inaccessible on port 8000 but worked on the random port.
And I found that using 127.0.0.1 or localhost is not the same thing in WSL.
I never paid attention to that before

**Comment by maMykola** [comment]
It works for me after `php-fpm` installation

**Comment by R3dLantern** [comment]
> It works for me after `php-fpm` installation

@maMykola Do you receive a continuous stream of errors in the terminal?

**Comment by maMykola** [comment]
> > It works for me after `php-fpm` installation
> 
> @maMykola Do you receive a continuous stream of errors in the terminal?

nope, I'm on ubuntu 22.04

**Comment by evanh-shumakov** [comment]
I'm on Windows 11 WSL2 with Ubuntu 22.04 and Xdebug work only with Development Server `[PHP        ] [Thu Jan 13 14:46:53 2022] PHP 8.0.14 Development Server (http://127.0.0.1:62470) started`, also I can't get access to 127.0.0.1:8000 from Chrome in Windows only to localhost:8000 or (again) to Development Server as it is. But the port is different every time, so I can't get a permanent settings

**Comment by evanh-shumakov** [comment]
It works after **php-fpm** is installed

**Comment by R3dLantern** [comment]
> > > It works for me after `php-fpm` installation
> > 
> > 
> > @maMykola Do you receive a continuous stream of errors in the terminal?
> 
> nope, I'm on ubuntu 22.04

But you use WSL, right?

**Comment by Nsbx** [comment]
The best solution for me was to add my WSL ip address in my host file and use this domain instead of localhost
![image](https://github.com/symfony-cli/symfony-cli/assets/8930930/2702163c-a04b-4d18-85c6-fbcad14bcf63)

**Comment by R3dLantern** [comment]
> It works after **php-fpm** is installed (Symfony 6, PHP 8.2, PHPStorm, Ubuntu 22.04, WSL 2, Windows 11)

Interesting - I guess WSL1 is the bottleneck then. Are your logs spammed by PHP-FPM?

**Comment by evanh-shumakov** [comment]
> > It works after **php-fpm** is installed (Symfony 6, PHP 8.2, PHPStorm, Ubuntu 22.04, WSL 2, Windows 11)
> 
> Interesting - I guess WSL1 is the bottleneck then. Are your logs spammed by PHP-FPM?

Not at all. You know, they have done great work with WSL2 last improvements. As I remember, when I first tried it I had to make some configurations just for my browser on Windows works with server on WSL2, now it works by default when you've installed it.

**Comment by tucksaun** [comment]
so after some digging, what is happening is that during the debugging session initialization PHPStorm sends the following eval command to Xdebug in order to retrieve the info`(string)($_SERVER['SERVER_PORT'])`

We do set `SERVER_PORT` to the port we listen to (https://github.com/symfony-cli/symfony-cli/blob/639fa67d7449c45addf616a6b1b327de5dc9757e/local/http/http.go#L234) but this ends up in `$_ENV` not `$_SERVER`, `$_SERVER['SERVER_PORT']` being set by the PHP webserver.

I tried to override  `$_SERVER['SERVER_PORT']` in the routing file but this file is actually executed *after* the session started so that does not work.

The next steps I want to try is:
- find a way to override $_SERVER natively (no big hopes here)
- try `autoprepend` and see if this one is executed before or not
- contact someone at Jetbrains to make them consider `$_ENV['SERVER_PORT']` before `$_SERVER['SERVER_PORT']`

**Comment by tucksaun** [comment]
After more research I didn't find a way to override `$_SERVER['SERVER_PORT']` early enough so that Phpstorm catches the right value.

@fabpot @nicolas-grekas do you have any contact at Jetbrains to whom we could suggest to check for `$_ENV['SERVER_PORT']` before `$_SERVER['SERVER_PORT']`?

**Comment by fabpot** [comment]
That would be @nicolas-grekas :)

**Comment by pronskiy** [comment]
Folks, sharing update from the PhpStorm team.

Checking $_ENV['SERVER_PORT'] does not appear to be a feasible approach. We would like to suggest an alternative solution: utilize the `PHP_IDE_CONFIG` env variable to define a serverName. This serverName should be consistent with the one configured in the _PHP | Servers_ settings. Doing so will enable us to establish the necessary path mappings for debugging connections. 

This technique is already effective for CLI debugging, and we would like to extend this approach to web debugging.

Does this solution make sense to you? If yes, we'll proceed with implementation.

**Comment by tucksaun** [comment]
@pronskiy thank you for the update and the time you put into helping us to solve this 🙏 

I will need to have a look, but because this might work 🙂
I will let you know later this week.

**Comment by Matt-PMCT** [comment]
@pronskiy Either I did something wrong or setting the ENV didn't work for me. I have server set in PHPSTORM:
![image](https://github.com/symfony-cli/symfony-cli/assets/6832191/5fd0dfb8-e819-46b7-a23c-1c86fa429d66)
Then I set a Linux ENV:
![image](https://github.com/symfony-cli/symfony-cli/assets/6832191/2ba32744-e7b9-4e25-b9c8-fc43fadc3ca7)
Then when using the SymfonyCLI server I still get incoming connection requests at the dev-port instead of port 8000 and that ENV variable didn't change that at all.

**Comment by LolGleb** [comment]
Greetings! The PhpStorm team has released a fix for this issue. It's been available since version 2024.1. I've tested it, and it seems to work fine. However, since there may be various project configurations, it would be great if you could verify if it works for you. I'll be looking forward to hearing from you.
